misc/printk.c

```diff
@@ -240,31 +241,49 @@ static int char_out(int c, struct out_context *ctx)
 
 	return c;
 }
-#else
+#endif /* CONFIG_USERSPACE */
+
 struct out_context {
 	int count;
 };
 
-static int char_out(int c, struct out_context *ctx)
+static int char_out(int c, void *ctx_p)
 {
+	struct out_context *ctx = (struct out_context *)ctx_p;
 	ctx->count++;
 	return _char_out(c);
 }
-#endif
 
+#ifdef CONFIG_USERSPACE
 int vprintk(const char *fmt, va_list ap)
 {
-	struct out_context ctx = { 0 };
+	if (_is_user_context()) {
```

Comment:
I don't have an alternative to this, but I wonder if we could do better than sprinkle calls to `_is_user_context()` in code that's essentially a library that kernel and userland can use.

Reply:
thinking about this.
was going to attack it when i revive the runtime library patch.
I do think printk() is a special case here, I don't expect to see a lot of this in library code. but it would be nice to omit the checks if __ZEPHYR_SUPERVISOR__ or __ZEPHYR_USER__ is defined